Homework 03

mrmegapolys opened the pull request.

Commit:
removed obsolete getters

Commit:
fixed bug with not closing file

Commit:
implemented alarm

Commit:
added alarm unittests

Commit:
security fix

Commit:
switched to another config file

Commit:
security fix

Commit:
added alarm to smart home and smart home custom serializer/deserializer

Commit:
refactoring exception handling in smartHomeSerializer

Commit:
changed arch of sensor event

Commit:
added alarmEvent and alarmEvent processor

Commit:
refactored supported events verification

Commit:
added notifier

Commit:
added activatedAlarm decorator

Commit:
cleanup & access modifiers fix

Commit:
decorators refactoring

mrmegapolys commented:
Извиняюсь за задержку (:

mrmegapolys commented:
У меня есть вопрос по поводу декораторов.

Изначально я сделал один декоратор, который совмещал функции тех двух, которые есть сейчас. Там был баг с двойным отправлением смски на DoorEvent, связано это с тем, что и SetDoorState и TurnLightsOff считают DoorEvent валидным для себя событием. 
Усиление проверки на валидность ничего не даст, ведь даже если мы не будем пускать в TurnLightsOff события, в которых фигурирует дверь не из холла, то на двери из холла все равно будет двойное срабатывание.

Первое решение, которое я увидел, это сделать так, чтобы был обработчик всех_дверей_кроме_той_что_в_холле, и обработчик двери в холле, который будет и обновлять состояние, и выключать свет. Так мы избегаем дублирования, но имхо в остальном мы сильно портим читаемость кода, однотипные действия теперь делаются в разных обработчиках. Плюс если способ изменения состояния двери изменится, нам придется править оба этих обработчика.

Другое решение это разнести декораторы, что я и сделал. Теперь проблема в конфигурационном коде. Я накидываю Notifier только на SetDoorEvent, потому что знаю, что SetDoorEvent и TurnLightsOff реагируют на одни и те же события. Соответственно, если вдруг возникнет необходимость выключить, к примеру, SetDoorEvent, то человек, не обладающий этим знанием, не добавит Notifier к TurnLightsOff, тем самым оставив нас без уведомлений.

Мне показалось, что второй вариант несколько лучше, тем не менее, он мне тоже не нравится. Что Вы считаете более правильным, а главное, нет ли третьего, менее спорного решения?
@a-ivanov @sbt-agoshkov

iagoshkov commented:
У нас по ТЗ, насколько я помню,  сообщение нужно отправлять в случае возникновения любого события в доме. Неважно будет оно отработано каким либо eventProcessor-ом или нет. Если принять это во внимание, то можно notifier сделать просто обычным обработчиком события,  а не декоратором. Чтобы он срабатывал один раз на каждое событие.

Если же хочется уведомлять только про те события, которые потенциально были бы обработаны eventProcessor-ами, то можно сделать отдельный класс, который брал бы на себя функции "дедупликации" сообщений. До передачи их в Notifier. Например, сравнивать какой-нибудь хэш от текущего события с уже отправленными уведомлениями и если хэш совпадает, то не уведомлять. Ну и тогда оставить декораторы. И из каждого декоратора передавать событие и сообщение в дедупликатор, который уже будет решать - отправлять или пропускать.

Commit:
notification sending refactoring

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventprocessors/EventProcessor.java:
```diff
@@ -1,7 +1,9 @@
 public interface EventProcessor {
     void process(SensorEvent event);
+
+    boolean isSupportedEvent(SensorEvent event);
```
Leaky abstraction - это деталь реализации, ее не должно быть в интерфейсе. Исправьте.

iagoshkov replied:
В домашке написано, что если сигнализация находится в состоянии "активирована", то любое событие от сенсоров должно переводить ее в режим тревоги - неважно есть на эти события eventProcessor-ы, которые собираются их обрабатывать или нет.
Поэтому это условие можно убрать из декоратора и убрать из интерфейса.

mrmegapolys replied:
Исправил

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventprocessors/decorators/ActivatedAlarmDecorator.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public void process(SensorEvent event) {
+        if (!isSupportedEvent(event)) return;
+        AlarmState state = alarm.getState();
```
А может Alarm быть Actionable? Она также находится в доме. Если да, то с ней можно работать так же, как с другими объектами дома.

iagoshkov replied:
Это замечание сняли

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/eventprocessors/processors/SetAlarmState.java:
```diff
@@ -0,0 +1,46 @@
+        int code = ((AlarmEvent) event).getCode();
+
+        if (event.getActionType() == AlarmActionType.ACTIVATE) {
+            activateAlarm(code);
```
См. замечания по Alarm как часть дома.

iagoshkov replied:
Это тоже сняли

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/smarthome/devices/alarm/Alarm.java:
```diff
@@ -0,0 +1,38 @@
+        this.state = state;
+    }
+
+    boolean isCorrectCode(int code) {
```
Отдельная responsibility, исправьте.

iagoshkov replied:
Решили, что это норм, можно так оставить. Замечание снято.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/smarthome/devices/alarm/Alert.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public void triggerAlert() {
+    }
```
Здесь можно писать в консоль оповещение, что сигнализация издает звук.

mrmegapolys replied:
Не очень понял где именно надо писать. Если в отмеченном участке кода, то смысла в этом особого не будет, имхо: этот метод же ничего не делает.
Может лучше написать в конструкторе класса Alert? Тогда оповещение будет один раз при срабатывании сигнализации. 
@a-ivanov

a-ivanov replied:
Это немного вопрос вкуса, но как мне кажется, когда явно вызывают triggerAlert, то как бы сигнализация звучит и это можно, но не обязательно, отразить, например сообщением в System.out.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/smarthome/devices/alarm/Deactivated.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public void activate(int code) {
+        if (alarm.isCorrectCode(code)) {
```
В этом состоянии задается изначальный код, его еще не с чем сверять.

iagoshkov replied:
Ну да, это типа как сейф в отеле. При активации ты задаешь код - потом этот же код должен использоваться для активации. Поэтому при активации код сравнивать не с чем, да и не нужно.

mrmegapolys replied:
Ну вообще в реальной жизни у домашней сигнализации код задается при ее установке и один раз...
Но ок, раз тз такое, переделаю

@a-ivanov @sbt-agoshkov

mrmegapolys replied:
Кстати, а что тогда делать с Alert-ом? Если сработал алерт когда сигнализация была активирована, то должна ли быть возможность выключить его тем же кодом, которым ее включали?

@sbt-agoshkov @a-ivanov

mrmegapolys replied:
А как тогда выключать алерт, который мы руками включили из Deactivated? Там не будет кода.
@sbt-agoshkov

mrmegapolys replied:
Исправил

iagoshkov commented on src/main/java/ru/sbt/mipt/oop/eventprocessors/processors/SetAlarmState.java:
```diff
@@ -0,0 +1,46 @@
+import ru.sbt.mipt.oop.smarthome.devices.alarm.AlarmActionType;
+import ru.sbt.mipt.oop.smarthome.devices.alarm.AlarmEvent;
+
+public class SetAlarmState implements EventProcessor {
```
Название этого (и других EventProcessor-ов) класса очень неудачное.
Классы обычно называются существительными, а не глаголами (глаголами - методы классов).
Лучше назвать, например, AlarmStateEventProcessor/DoorStateEventPorcessor/LightStateEventProcessor.

mrmegapolys replied:
Исправил

Commit:
removed isSupportedEvent

Commit:
rename of event processors

iagoshkov commented:
Должна20:05, 3 ноября 2019 г., mrmegapolys <notifications@github.com>:@mrmegapolys commented on this pull request.



In src/main/java/ru/sbt/mipt/oop/smarthome/devices/alarm/Deactivated.java:
> @@ -0,0 +1,25 @@
+package ru.sbt.mipt.oop.smarthome.devices.alarm;
+
+public class Deactivated implements AlarmState {
+    transient private final Alarm alarm;
+
+    Deactivated(Alarm alarm) {
+        this.alarm = alarm;
+    }
+
+    @Override
+    public void activate(int code) {
+        if (alarm.isCorrectCode(code)) {

Кстати, а что тогда делать с Alert-ом? Если сработал алерт когда сигнализация была активирована, то должна ли быть возможность выключить его тем же кодом, которым ее включали?
@sbt-agoshkov @a-ivanov

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
Илья

iagoshkov commented:
Хороший вопрос ))))Тогда наверно не надо выключать ))20:10, 3 ноября 2019 г., mrmegapolys <notifications@github.com>:@mrmegapolys commented on this pull request.



In src/main/java/ru/sbt/mipt/oop/smarthome/devices/alarm/Deactivated.java:
> @@ -0,0 +1,25 @@
+package ru.sbt.mipt.oop.smarthome.devices.alarm;
+
+public class Deactivated implements AlarmState {
+    transient private final Alarm alarm;
+
+    Deactivated(Alarm alarm) {
+        this.alarm = alarm;
+    }
+
+    @Override
+    public void activate(int code) {
+        if (alarm.isCorrectCode(code)) {

А как тогда выключать алерт, который мы руками включили из Deactivated? Там не будет кода.
@sbt-agoshkov

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
Илья

Commit:
made alarm code String instead of int

Commit:
Alarm code is now set on activation instead of alarm creation.
Alert triggered from Deactivated can be turned off with default code,
which is set on alarm creation.
